contrib/meson/meson/programs/meson.build

```diff
@@ -1,52 +1,61 @@
 # #############################################################################
-# Copyright (c) 2018-present    lzutao <taolzu(at)gmail.com>
+# Copyright (c) 2018-present        lzutao <taolzu(at)gmail.com>
+# Copyright (c) 2022-present        Tristan Partin <tristan(at)partin.io>
 # All rights reserved.
 #
 # This source code is licensed under both the BSD-style license (found in the
 # LICENSE file in the root directory of this source tree) and the GPLv2 (found
 # in the COPYING file in the root directory of this source tree).
 # #############################################################################
 
-lz4_root_dir = '../../../..'
-
-lz4_includes = include_directories(join_paths(lz4_root_dir, 'programs'))
-lz4_sources = [join_paths(lz4_root_dir, 'programs/bench.c'),
-  join_paths(lz4_root_dir, 'programs/datagen.c'),
-  join_paths(lz4_root_dir, 'programs/lz4cli.c'),
-  join_paths(lz4_root_dir, 'programs/lz4io.c')]
-lz4_c_args = []
-
-export_dynamic_on_windows = false
-# explicit backtrace enable/disable for Linux & Darwin
-if not use_backtrace
-  lz4_c_args += '-DBACKTRACE_ENABLE=0'
-elif use_debug and host_machine_os == os_windows  # MinGW target
-  lz4_c_args += '-DBACKTRACE_ENABLE=1'
-  export_dynamic_on_windows = true
+sources = files(
+  lz4_source_root / 'programs/bench.c',
+  lz4_source_root / 'programs/datagen.c',
+  lz4_source_root / 'programs/lz4cli.c',
+  lz4_source_root / 'programs/lz4io.c',
+)
+
+lz4 = executable(
+  'lz4',
+  sources,
+  objects: liblz4.extract_all_objects(recursive: true),
```

Comment:
Why do you need to extract all objects here? Guessing because the build never used gnu symbol visibility before, *and* the executable is relying on symbols that aren't exported with LZ4LIB_API?

Reply:
Exactly. @Cyan4973 any thoughts on this?

Reply:
I don't remember the details, but I know that the CLI executable is designed to be linked to the _static_ variant of the library, meaning it is not restricted to "public" symbols. One such example is the `xxhash` library, which is not exposed on the public side of `liblz4`, but is employed internally to calculate checksums. The same function is used either as a random number generator or as a checksum for different purposes within the CLI.

Reply:
The most elegant solution here, IMHO, is to do a bit of project restucturing until you end up with:
- `libcommon.a`: a static library containing internal symbols shared between the library and the CLI
- `liblz4.so`/`liblz4.a`: a shared/static library containing public symbols of the library only. It links to libcommon
- `lz4`: a CLI executable that links to both libcommon and to liblz4

So, the CLI executable would be free to link to the shared variant of the library, and it ***is*** restricted to public symbols, but any time you decide you need a private symbol, you just move the private symbol to libcommon.

The CLI executable then prefers to link to public symbols from liblz4, which depending on the vendor's choice may be static or shared, because some vendors have decided for some reason they like shared libraries ;)

But, for any non-public symbols, the CLI executable just falls back to statically linking them from the convenience library `libcommon`.

You get the best of both worlds. The downside is it requires more work.

Reply:
I want to copy essentially what upstream is currently doing. So for me I see two options other than the re-org suggested by @eli-schwartz. Link against `liblz4_internal_dep` which is an internal static library or continue to do what I am doing currently which is just extracting objects from the static library.